src/rdkafka_sticky_assignor.c

```diff
@@ -25,170 +25,3465 @@
+        }
+
+        RD_MAP_DESTROY(&preBalancePartitionConsumers);
+        RD_MAP_DESTROY(&preBalanceAssignment);
+
+        /* Add the fixed assignments (those that could not change) back. */
+        if (!RD_MAP_IS_EMPTY(&fixedAssignments)) {
+                const rd_map_elem_t *elem;
+
+                RD_MAP_FOREACH_ELEM(elem, &fixedAssignments.rmap) {
+                        const char *consumer = elem->key;
+                        rd_kafka_topic_partition_list_t *partitions =
+                                (rd_kafka_topic_partition_list_t *)elem->value;
+
+                        RD_MAP_SET(currentAssignment, consumer, partitions);
+
+                        rd_list_add(sortedCurrentSubscriptions, (void *)elem);
+                }
+
+                /* Re-sort */
+                rd_list_sort(sortedCurrentSubscriptions,
+                             sort_by_map_elem_val_toppar_list_cnt);
+        }
+
+        RD_MAP_DESTROY(&fixedAssignments);
+}
+
+
+
+
+
+
+
+
+
+
+/**
+ * @brief Populate subscriptions, current and previous assignments based on the
+ *        \p members assignments.
+ */
+static void
+preopulateCurrentAssignments (
+        rd_kafka_t *rk,
+        rd_kafka_group_member_t *members,
+        size_t member_cnt,
+        map_str_toppar_list_t *subscriptions,
+        map_str_toppar_list_t *currentAssignment,
+        map_toppar_cgpair_t *prevAssignment,
+        map_toppar_str_t *currentPartitionConsumer,
+        map_str_toppar_list_t *consumer2AllPotentialPartitions,
+        size_t estimated_partition_cnt) {
+
+        /* We need to process subscriptions' user data with each consumer's
+         * reported generation in mind.
+         * Higher generations overwrite lower generations in case of a conflict.
+         * Conflicts will only exist if user data is for different generations.
+         */
+
+        /* For each partition we create a sorted list (by generation) of
+         * its consumers. */
+        RD_MAP_LOCAL_INITIALIZER(sortedPartitionConsumersByGeneration,
+                                 member_cnt * 10 /* FIXME */,
+                                 const rd_kafka_topic_partition_t *,
+                                 /* List of ConsumerGenerationPair_t */
+                                 rd_list_t *,
+                                 rd_kafka_topic_partition_cmp,
+                                 rd_kafka_topic_partition_hash,
+                                 NULL,
+                                 rd_list_destroy_free);
+        const rd_kafka_topic_partition_t *partition;
+        rd_list_t *consumers;
+        int i;
+
+        /* For each partition that is currently assigned to the group members
+         * add the member and its generation to
+         * sortedPartitionConsumersByGeneration (which is sorted afterwards)
+         * indexed by the partition. */
+        for (i = 0 ; i < (int)member_cnt ; i++) {
+                rd_kafka_group_member_t *consumer = &members[i];
```

Comment:
`member` seems a better name here than `consumer`, since that's what they're typically referred to in this context. i think consumer is less clear.

Reply:
This is using the Java implementation variable names to make code maintenance easier.

member and consumer is interchangable in th is context.